src/Frontpage/Frontpage.js

```diff
@@ -1,17 +1,75 @@
 import * as React from 'react';
 import './Frontpage.css';
 import DisplayDate from '../DisplayDate/DisplayDate'
+import CanClass from '../CanvasClass/CanClass';
 
 const Frontpage = (props) => {
+  const players= [
+    {name: 'Thomas'},
+    {name: 'Boushra'},
+    {name: 'Gayathri'},
+    {name: 'Tahir'},
+    {name: 'Vinh'},
+    {name: 'Ingar'},
+    {name: 'Mary'},
+    {name: 'Irlend'},
+    {name: 'Maha'},
+    {name: 'Yusuf'}
+  ]
+    
+  const showPlayer = () =>{
+    let i=0
+    let index=0
+    let canvas_list=players.map(player=><CanClass playerName={player.name} key={player.name}/>)
+    let players_show_list=[]
+    //console.log(players.length)
+    
+    if(players.length%3===0){
+      while(i<(players.length/3))
+      {console.log(players.length)
+        players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td><td>{canvas_list[index+1]}</td><td>{canvas_list[index+2]}</td></tr>)
+        index=index+3
+        i=i+1
+      }
+    }
+    
+    else if(players.length%3===1){
+      while(i<(players.length/3))
+      {
+        players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td><td>{canvas_list[index+1]}</td><td>{canvas_list[index+2]}</td></tr>)
+        index=index+3
+        i++
+      }
+        players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td></tr>)
+    }
+    else
+    {    while(i<(players.length/3))
+      {
+        players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td><td>{canvas_list[index+1]}</td><td>{canvas_list[index+2]}</td></tr>)
+        index=index+3
+        i++
+      }
+      players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td><td>{canvas_list[index+1]}</td></tr>)
+   }
+
+    return(
+    <tbody>
+      {players_show_list}
+    </tbody>
+    )
+  }
```

Comment:
<h2> Your code works, great job!  </h2>
I want to share some thoughts with you about how you wrote this code. You don't need to change anything, but I think it can be useful to go through this, as it will show you some better practices for writing code. <br/>
First of all it was hard for me to read what was happening. Ideally you want to make your code as readable as possible, as when you start working in tech, a lot of your time will be spent reading others peoples code. I also think if you come back to this code after a few weeks (this has happened to be many times), you will have trouble understanding what you've done here as well.

## How to improve your codes readability
### Better formatting
Formatting used to be a heated topic among developers. It's very important, as good formatting makes it easier to read the code and understand the scope and logic. The best way to solve this is to use  [https://prettier.io/](Prettier) (I can help you install it next session.) Prettier automatically formats your code, so you don't have to worry about it.
### Separate render logic and data manipulation
This problem is called **Separation of Concerns**. It is good practice to always try to separate code that does different things. <br/>
<h3>Example </h3>

``` 
if(players.length%3===0){
      while(i<(players.length/3))
      {console.log(players.length)
        players_show_list.push(<tr key={i}><td>{canvas_list[index]}</td><td>{canvas_list[index+1]}</td><td>{canvas_list[index+2]}</td></tr>)
        index=index+3
        i=i+1
      }
    } 
```

Here you handle both the **data manipulation** part of the code (your _if_ and _while_ statements) and the **user interface** (what renders) in the same block. This is not ideal, because if you wanted to come back and change the code, say, you want divide the group in 4's not 3's. You would have to go and make changes in this block and the two other blocks you have.

Now I just want to remind you, we don't expect you to know about all these best practices. But for the educational purpose I re-wrote your code so you could see one way you could to this.

```
const displayPlayersInGroup = (playersPerGroup, players) => {
  const arrayOfGroups = [];
  let chunkOfPlayers = [];

  players.map(player => {
    chunkOfPlayers.push(player);
    if (chunkOfPlayers.length === playersPerGroup) {
      arrayOfGroups.push(chunkOfPlayers);
      chunkOfPlayers = [];
    }
  });
  arrayOfGroups.push(chunkOfPlayers);

  return arrayOfGroups.map((group, i) => (
    <tr key={i}>
      {group.map(player => (
        <td>
          <CanvasClass playerName={player.name} key={player.name} />
        </td>
      ))}
    </tr>
  ));
};
```

This is by no means the best way to write this code. There are more things we could do like separating the data manipulation into it's own function. But this is a good start.

This code might be difficult for you to read. I encourage you to still try. 
The function receives two parameters. **playersPerGroup** which is how many players there should be in a group. And **players** which is an array with all the players.

Here is the code again with some added comments to help you understand what's happening:
```
const displayPlayersInGroup = (playersPerGroup, players) => {
  const arrayOfGroups = [];
  let chunkOfPlayers = [];


//comment: here I handle the logic 👇🏻 (a for loop might be better suited than map for this. I wasn't sure if you guys had used them)
  players.map(player => {
    chunkOfPlayers.push(player);
//comment: I keep adding to chuckOfPlayers until the length is equal to the amount playersPerGroup.
    if (chunkOfPlayers.length === playersPerGroup) {
      arrayOfGroups.push(chunkOfPlayers);
//comment: When the amount is equal, i push the array with players to arrayOfGroups and reset chunkOfPlayers til and empty array.
      chunkOfPlayers = [];
    }
  });
//comment: Here I add the leftover players if there are any.
  arrayOfGroups.push(chunkOfPlayers);

// comment: the array of groups looks something like this in the end: [ [player1, player2, player3], [player4, player5, player6], [player7, player8, player9] ]

//comment:  here I render the tables 👇🏻
  return arrayOfGroups.map((group, i) => (
    <tr key={i}>
      {group.map(player => (
        <td>
          <CanvasClass playerName={player.name} key={player.name} />
        </td>
      ))}
    </tr>
  ));
};
```

I hope you find this useful, and it doesn't just read as a long ramble from me 😅 tell me if there is something unclear!

Reply:
Thank you for sharing that. Yes, actually I do have a problem with the points that you've mentioned due to being out of the game for four years now :( , so I sort of forgot how to write a 'clean code'. I'll try to enhance my code next time 🤞.